Overrides rewrite in PointRangeQuery to optimize AllDocs/NoDocs cases

ebradshaw opened the pull request:
Overrides rewrite in PointRangeQuery range to handle cases where the query either fully contains or fully excludes all documents within the shard.

Often, particularly when using time based partitioning, range queries may overlap several indexes.  Many of these indexes have timestamp values that are fully contained by the query, in which case the query can be rewritten to a MatchAllDocsQuery or a FieldExistsQuery.  On the other hand, many indexes can be fully excluded if they're outside the requested time range, in which case the query can be rewritten to a MatchNoDocsQuery.

While a similar optimization exists at the leaf level in the createWeight function, rewriting at the shard level enables other optimizations downstream.

Please let me know if this has been ruled out in the past for other reasons or if the implementation misses anything.  Thanks.

ebradshaw committed:
Overrides rewrite in PointRangeQuery to handle cases where the query range either fully contains or fully excludes all documents within the shard.

ebradshaw committed:
Fixes wildcarded imports

ebradshaw committed:
Null check on values

jpountz commented:
+1 to do this

Feel free to ask me for a review on a PR if you'd live to give it a try.

jpountz reviewed (commented):
I left some minor comments but the change looks good to me in general. Thank you!

jainankitk commented on lucene/core/src/java/org/apache/lucene/search/PointRangeQuery.java:
```diff
@@ -580,4 +557,111 @@ public final String toString(String field) {
+        fullyContainedCount++;
+      } else if (qLow.compareTo(gMax) > 0 || qHigh.compareTo(gMin) < 0) {
+        fullyExcludedCount++;
+      }
```
We can `return super.rewrite(searcher)` if any dimension is not fullyContained / Excluded?

ebradshaw replied:
Since we're using the relate method, we'll get the short circuit for excluded, but it'll still run through every contained check to determine inside vs crosses.  Should be okay since it's just a rewrite call?

jainankitk replied:
Yeah, as you said this is only for rewrite and logic is much cleaner by reusing the relate method

jainankitk commented on lucene/core/src/java/org/apache/lucene/search/PointRangeQuery.java:
```diff
@@ -580,4 +557,111 @@ public final String toString(String field) {
+      }
+    } else if (fullyExcludedCount == dims) {
+      return new MatchNoDocsQuery();
+    }
```
+1 to using `relate(byte[], byte[])`. Actually looking at `relate(byte[], byte[])` method, I realized that we don't need `fullyExcludedCount == dims` for rewriting into `MatchNoDocsQuery`. Any dimension being completely outside is sufficient condition for the query to be rewritten as `MatchNoDocsQuery`

ebradshaw replied:
Agreed.  Missed the relate method on the first run but that logic is nice and clean.

ebradshaw committed:
Addresses MR comments:

- Fixes annotation ordering
- Pulls relate method out of inner class and refactors rewrite to use it

ebradshaw commented:
I think we're good now.  Let me know if you guys see anything else, and thanks for the quick review!

jpountz commented on lucene/core/src/java/org/apache/lucene/search/PointRangeQuery.java:
```diff
@@ -580,4 +534,114 @@ public final String toString(String field) {
+    byte[] globalMaxPacked = PointValues.getMaxPackedValue(reader, getField());
+
+    if (globalMinPacked == null || globalMaxPacked == null) {
+      return super.rewrite(searcher);
```
A null min or max packed value indicates an empty field, so you could rewrite to a `MatchNoDocsQuery` here as well.

jpountz reviewed (approved):
LGTM, can you add a CHANGES entry?

ebradshaw committed:
Adds CHANGES.txt entry and updates PointRangeQuery.rewrite to return MatchNoDocsQuery if no point values exist

jainankitk commented:
Thanks @ebradshaw for this change, and congratulations on your first contribution in lucene! :)
